Add validating webhook for shoots

timuthy opened the pull request:
**What this PR does / why we need it**:
This PR adds a validating webhook that is responsible for checking resources of the group `core.gardener.cloud` (at the moment only `shoot` resources).

With [GEP-4](https://github.com/gardener/gardener/blob/master/docs/proposals/04-new-core-gardener-cloud-apis.md) cloud specific information was externalized and as a consequence cannot be validated by the Gardener anymore. On the other hand, it should be prevented that invalid `shoot` objects are created or updated.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy|action
- target_group:   user|operator|developer
-->
```noteworthy operator
The `gardener-extension-provider-azure` now offers a validating webhook which checks shoot resources of type `azure` in the `core.gardener.cloud` group. Operators should register this webhook in the Garden cluster to further prevent invalid modifications on `azure` shoots.
```

rfranzke commented on charts/gardener-extension-validator-azure/charts/application/Chart.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: A Helm chart to deploy the gardener-extensions-validator-azure application related resources
```
```suggestion
description: A Helm chart to deploy the gardener-extension-validator-azure application related resources
```

rfranzke commented on charts/gardener-extension-validator-azure/charts/application/templates/_helpers.tpl:
```diff
@@ -0,0 +1,23 @@
+{{- define "name" -}}
+gardener-extensions-validator-azure
```
```suggestion
gardener-extension-validator-azure
```

rfranzke replied:
FYI: We should also rename AWS...

rfranzke replied:
But I guess this should be a symlink to `../../templates/helpers.tpl`?

rfranzke commented on charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml:
```diff
@@ -0,0 +1,30 @@
+  namespaceSelector: {}
+  clientConfig:
+    {{- if .Values.global.virtualGarden.enabled }}
+    url: {{ printf "https://gardener-extensions-validator-azure.%s/webhooks/validate-shoot" .Release.Namespace }}
```
```suggestion
    url: {{ printf "https://gardener-extension-validator-azure.%s/webhooks/validate-shoot" .Release.Namespace }}
```

rfranzke commented on charts/gardener-extension-validator-azure/charts/runtime/Chart.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: A Helm chart to deploy the gardener-extensions-validator-azure runtime related resources
```
```suggestion
description: A Helm chart to deploy the gardener-extension-validator-azure runtime related resources
```

rfranzke commented on charts/gardener-extension-validator-azure/charts/application/values.yaml:
```diff
@@ -0,0 +1,26 @@
+global:
```
This should be a symlink to `../../values.yaml`

rfranzke commented on charts/gardener-extension-validator-azure/charts/runtime/templates/_helpers.tpl:
```diff
@@ -0,0 +1,23 @@
+{{- define "name" -}}
```
symlink?

rfranzke commented on charts/gardener-extension-validator-azure/charts/runtime/templates/deployment.yaml:
```diff
@@ -0,0 +1,67 @@
+  template:
+    metadata:
+      annotations:
+        checksum/secret-gardener-extensions-validator-azure-cert: {{ include (print $.Template.BasePath "/secret-cert.yaml") . | sha256sum }}
```
s/extensions/extension everywhere

rfranzke commented on charts/gardener-extension-validator-azure/charts/runtime/templates/secret-cert.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gardener-extensions-validator-azure-cert
```
```suggestion
  name: gardener-extension-validator-azure-cert
```

rfranzke commented on charts/gardener-extension-validator-azure/charts/runtime/templates/secret-kubeconfig.yaml:
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gardener-extensions-validator-azure-kubeconfig
```
```suggestion
  name: gardener-extension-validator-azure-kubeconfig
```

rfranzke commented on charts/gardener-extension-validator-azure/charts/runtime/values.yaml:
```diff
@@ -0,0 +1,26 @@
+global:
```
This should be a symlink to `../../values.yaml`

rfranzke commented on charts/gardener-extension-validator-azure/templates/_helpers.tpl:
```diff
@@ -0,0 +1,23 @@
+{{- define "name" -}}
+gardener-extensions-validator-azure
```
```suggestion
gardener-extension-validator-azure
```

rfranzke commented on cmd/gardener-extension-validator-azure/app/app.go:
```diff
@@ -0,0 +1,83 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var log = logf.Log.WithName("gardener-extensions-validator-azure")
```
```suggestion
var log = logf.Log.WithName("gardener-extension-validator-azure")
```

rfranzke commented on example/40-validatingwebhookconfiguration.yaml:
```diff
@@ -0,0 +1,23 @@
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: gardener-extensions-validator-azure
```
```suggestion
  name: gardener-extension-validator-azure
```

rfranzke commented on example/validator-azure-certs/tls.crt:
```diff
@@ -0,0 +1,23 @@
+-----BEGIN CERTIFICATE-----
```
It doesn't harm, but this cert contains "Provider AWS" which might be confusing.

rfranzke commented on pkg/apis/azure/validation/infrastructure.go:
```diff
@@ -48,13 +48,10 @@ func ValidateInfrastructureConfig(infra *apisazure.InfrastructureConfig, resourc
 	}
 
-	networksPath := field.NewPath("networks")
-	if len(infra.Networks.Workers) == 0 {
```
Why was this removed?

timuthy replied:
`cidrvalidation.NewCIDR` in line 56 already validates the correct CIDR range and an empty string is also invalid. We can also do it explicitly if you if you prefer the way it was done before?

timuthy replied:
The point is that before the user got two error messages:

```
              {
                  Type: "FieldValueRequired",
                  Field: "networks.workers",
                  BadValue: "",
                  Detail: "must specify the network range for the worker network",
              },
              {
                  Type: "FieldValueInvalid",
                  Field: "networks.workers",
                  BadValue: "",
                  Detail: "invalid CIDR address: ",
              },
```

So I thought it would be sufficient to only return `invalid CIDR address`.

rfranzke commented on pkg/apis/azure/validation/infrastructure.go:
```diff
@@ -67,8 +64,8 @@ func ValidateInfrastructureConfig(infra *apisazure.InfrastructureConfig, resourc
 		}
-		if *infra.Networks.VNet.ResourceGroup == *resourceGroupName {
-			allErrs = append(allErrs, field.Invalid(networksPath.Child("vnet", "resourceGroup"), *infra.Networks.VNet.ResourceGroup, "specifying an existing vnet is the cluster resource group is not supported"))
+		if infra.ResourceGroup != nil && *infra.Networks.VNet.ResourceGroup == infra.ResourceGroup.Name {
```
Why have you added `infra.ResourceGroup != nil`? It's already in line 63

timuthy commented:
Thanks for the review @rfranzke. PTAL.

rfranzke commented on charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml:
```diff
@@ -0,0 +1,30 @@
+metadata:
+  name: {{ include "name" . }}
+webhooks:
+- name: shoot.azure.provider.extensions.gardener.cloud
```
```suggestion
- name: validation.azure.provider.extensions.gardener.cloud
```

rfranzke commented on charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml:
```diff
@@ -0,0 +1,30 @@
+    service:
+      namespace: {{ .Release.Namespace }}
+      name: {{ include "name" . }}
+      path: /webhooks/validate-shoot
```
```suggestion
      path: /webhooks/validate
```

rfranzke commented on example/40-validatingwebhookconfiguration.yaml:
```diff
@@ -0,0 +1,23 @@
+metadata:
+  name: gardener-extension-validator-azure
+webhooks:
+- name: shoot.azure.provider.extensions.gardener.cloud
```
```suggestion
- name: validation.azure.provider.extensions.gardener.cloud
```

rfranzke commented on example/40-validatingwebhookconfiguration.yaml:
```diff
@@ -0,0 +1,23 @@
+  objectSelector: {}
+  namespaceSelector: {}
+  clientConfig:
+    url: "https://localhost:9443/webhooks/validate-shoot"
```
```suggestion
    url: "https://localhost:9443/webhooks/validate"
```

zanetworker commented on pkg/validator/shoot_validator.go:
```diff
@@ -0,0 +1,113 @@
+	// Provider validation
+	fldPath := field.NewPath("spec", "provider")
+
+	if errList := azurevalidation.ValidateInfrastructureConfig(infraConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services, fldPath); len(errList) != 0 {
```
shouldn't this be `infraConfigFldPath` instead? 

```suggestion
	if errList := azurevalidation.ValidateInfrastructureConfig(infraConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services, infraConfigFldPath); len(errList) != 0 {
```

timuthy commented:
I addressed the review comments in the latest update. Furthermore, @tim-ebert suggested to aggregate the errors instead of early exit everywhere. This suggestion has been adapted where applicable.

rfranzke commented on pkg/validator/shoot_validator.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+func (v *Shoot) validateShootCreation(shoot *core.Shoot) error {
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
```
Why are lines 93-100 not moved into the `validateShoot` function as well? The `ControlPlaneConfig` related code is in this function as well.

timuthy replied:
You need to do it before because in `validateShootUpdate` also the "old" object needs to be decoded + checked. If it were in `validateShoot ` we'd do it twice.

rfranzke commented on pkg/validator/shoot_validator.go:
```diff
@@ -0,0 +1,103 @@
+	infraConfig, err := decodeInfrastructureConfig(v.decoder, shoot.Spec.Provider.InfrastructureConfig)
+	if err != nil {
+		return field.Forbidden(infraConfigPath, "not allowed to configure an unsupported infrastructureConfig")
+	}
```
Maybe it makes sense to combine this into a function as it is done multiple times (first `nil` check, then decode check).
